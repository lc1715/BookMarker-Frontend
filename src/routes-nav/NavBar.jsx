import { useState, useContext } from 'react';
import UserContext from '../auth/UserContext';
//Material UI
import AppBar from '@mui/material/AppBar';
import Box from '@mui/material/Box';
import Toolbar from '@mui/material/Toolbar';
import Button from '@mui/material/Button';
import '@fontsource/roboto/400.css';
import Avatar from '@mui/material/Avatar';
import Menu from '@mui/material/Menu';
import MenuItem from '@mui/material/MenuItem';
import IconButton from '@mui/material/IconButton';
import Tooltip from '@mui/material/Tooltip';
import Link from '@mui/material/Link';
import { useTheme } from '@mui/material/styles';
import useMediaQuery from '@mui/material/useMediaQuery';


/** Navigation bar for site. 
 * Shows up on every page.
 *
 * When user is logged in, shows links to main areas of site. 
 * When not logged in, shows link to Login and Signup forms.
 *
 * Rendered by App.jsx
 */

function NavBar({ logout }) {
    let { currentUser } = useContext(UserContext);
    const [anchorEl, setAnchorEl] = useState(null);
    //useMediaQuery hook
    const theme = useTheme();
    const matches = useMediaQuery(theme.breakpoints.up('md'));

    const handleOpenMenu = (evt) => {
        setAnchorEl(evt.currentTarget);
    }

    const handleCloseMenu = () => {
        setAnchorEl(null);
    }

    const handleLogout = () => {
        logout();
        setAnchorEl(null);
    }

    function loggedInNav() {
        return (
            <Box sx={{ flexGrow: 1 }}>
                <AppBar position="static" sx={{
                    backgroundColor: '#748cab',
                    boxShadow: 'none'
                }}>
                    <Toolbar sx={{ flexWrap: 'wrap', gap: 2, pl: 1, height: { lg: '69px' }, alignContent: { lg: 'center' } }} >
                        <Button href='/'
                            sx={{ color: 'white', textTransform: 'none', fontSize: { xs: '21px', sm: '18px' } }}
                            aria-label='Navigate to homepage'>
                            <i className="fa-solid fa-feather" style={{ marginRight: 6 }}></i>
                            BookMarker
                        </Button>

                        {/* On medium screens and larger, show full navbar */}
                        {matches ?
                            <>
                                <Button href='/books/read'
                                    sx={{ color: 'white', fontSize: 'medium' }}
                                    aria-label='Navigate to read books'>
                                    Read
                                    <i className="fa-solid fa-book" style={{ marginLeft: '4px' }}></i>
                                </Button>
                                <Button href='/books/wish'
                                    sx={{ color: 'white', fontSize: 'medium' }}
                                    aria-label='Navigate to wish to read books'>
                                    Wish To Read
                                    <i className="fa-solid fa-book-open" style={{ marginLeft: '4px' }}></i>
                                </Button>

                                {/* Takes up space between nav links */}
                                <Box sx={{ flexGrow: 1 }} />

                                <Button href='/profile' sx={{ color: 'white', my: 2, fontSize: 'medium' }}>
                                    Profile
                                </Button>
                                <Button href='/' onClick={logout} sx={{ color: 'white', my: 2, fontSize: 'medium' }}>
                                    LogOut
                                </Button>
                            </>
                            :
                            <>
                                {/* On smaller screens, show user menu*/}
                                <Box sx={{ flexGrow: 1 }} />

                                <Box sx={{ display: 'flex', alignItems: 'center', textAlign: 'center' }}>
                                    <Tooltip title="open-menu">
                                        <IconButton
                                            onClick={handleOpenMenu}
                                            size="large"
                                            sx={{ py: 0 }}
                                            aria-controls={open ? 'user-menu' : undefined}
                                            aria-haspopup="true"
                                            aria-expanded={open ? 'true' : undefined}
                                        >
                                            <Avatar sx={{ width: 41, height: 40 }}>
                                                {currentUser.username.slice(0, 1).toUpperCase()}
                                            </Avatar>
                                        </IconButton>
                                    </Tooltip>
                                </Box>

                                <Menu
                                    sx={{ mt: '4.5px' }}
                                    id="menu-list"
                                    open={Boolean(anchorEl)}
                                    onClose={handleCloseMenu}
                                    anchorEl={anchorEl}
                                    transformOrigin={{
                                        vertical: 'top',
                                        horizontal: 'right'
                                    }}
                                    anchorOrigin={{
                                        vertical: 'bottom',
                                        horizontal: 'right'
                                    }}
                                >
                                    <MenuItem onClick={handleCloseMenu}>
                                        <Link href="/books/read" underline="none" sx={{ fontSize: '18px' }}
                                            aria-label='Navigate to read books'>
                                            Read
                                            <i className="fa-solid fa-book" style={{ marginLeft: '5px' }}></i>
                                        </Link>
                                    </MenuItem>

                                    <MenuItem onClick={handleCloseMenu}>
                                        <Link href="/books/wish" underline="none" sx={{ fontSize: '18px' }}
                                            aria-label='Navigate to wish to read books'>
                                            Wish To Read
                                            <i className="fa-solid fa-book-open" style={{ marginLeft: '5px' }}></i>
                                        </Link>
                                    </MenuItem>

                                    <MenuItem onClick={handleCloseMenu}>
                                        <Link href="/profile" underline="none" sx={{ fontSize: '18px' }}>
                                            Profile
                                        </Link>
                                    </MenuItem>

                                    <MenuItem onClick={handleLogout}>
                                        <Link href="/" underline="none" sx={{ fontSize: '18px' }}>
                                            Logout
                                        </Link>
                                    </MenuItem>
                                </Menu>
                            </>
                        }
                    </Toolbar>
                </AppBar>
            </Box >
        )
    };

    function loggedOutNav() {
        return (
            <Box sx={{ flexGrow: 1 }}>
                <AppBar position="static" sx={{
                    backgroundColor: '#748cab',
                    boxShadow: 'none'
                }}>
                    <Toolbar sx={{ flexWrap: 'wrap', height: { lg: '69px' }, alignContent: { lg: 'center' } }}>
                        <Button href='/'
                            sx={{ pl: { xs: 0 }, textTransform: 'none', color: 'white', fontSize: '19px' }}
                            aria-label='Navigate to homepage'>
                            <i className="fa-solid fa-feather" style={{ marginRight: 6 }}></i>
                            BookMarker
                        </Button>

                        <Box sx={{ flexGrow: 1 }} />

                        <Button href='/login' sx={{ mr: 1, my: 2, color: 'white', fontSize: { xs: '15px', sm: 'medium' } }}>
                            Login
                        </Button>
                        <Button href='/signup' sx={{ pl: 0, color: 'white', fontSize: { xs: '15px', sm: 'medium' } }}>
                            Sign Up
                        </Button>
                    </Toolbar>
                </AppBar>
            </Box >
        )
    };

    return (
        <>
            {currentUser ? loggedInNav() : loggedOutNav()}
        </>
    );
};

export default NavBar;
