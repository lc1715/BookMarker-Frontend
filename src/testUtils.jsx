import UserContext from "./auth/UserContext";

const demoUser = {
    id: 1,
    username: 'testUser',
    email: 'test@email.com',
    volume_ids: ['e3', 't5'],
}

const UserProvider =
    ({ children, currentUser = demoUser, hasSavedBook = () => (true) }) => (
        <UserContext.Provider value={{ currentUser, hasSavedBook }}>
            {children}
        </UserContext.Provider>
    );

export { UserProvider };
