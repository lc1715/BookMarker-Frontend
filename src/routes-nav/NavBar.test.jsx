import React from "react";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router";
import NavBar from "./NavBar";
import { UserProvider } from "../testUtils";

it('renders the NavBar component', () => {
    render(
        <MemoryRouter>
            <UserProvider>
                <NavBar />
            </UserProvider>
        </MemoryRouter>
    )
})

it('matches with snapshot when logged in', async () => {
    const { asFragment } = render(
        <MemoryRouter>
            <UserProvider>
                <NavBar />
            </UserProvider>
        </MemoryRouter>
    )

    expect(asFragment()).toMatchSnapshot();
})

it("matches with snapshot when logged out", function () {
    const { asFragment } = render(
        <MemoryRouter>
            <UserProvider currentUser={null}>
                <NavBar />
            </UserProvider>
        </MemoryRouter>
    )

    expect(asFragment()).toMatchSnapshot();
});
