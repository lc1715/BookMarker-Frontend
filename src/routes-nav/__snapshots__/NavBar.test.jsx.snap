// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches with snapshot when logged in 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-i9gxme"
  >
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-1fha3o-MuiPaper-root-MuiAppBar-root"
      style="--Paper-shadow: 0px 2px 4px -1px rgba(0,0,0,0.2),0px 4px 5px 0px rgba(0,0,0,0.14),0px 1px 10px 0px rgba(0,0,0,0.12);"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-mywmkt-MuiToolbar-root"
      >
        <a
          aria-label="Navigate to homepage"
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-c1pv86-MuiButtonBase-root-MuiButton-root"
          href="/"
          tabindex="0"
        >
          <i
            class="fa-solid fa-feather"
            style="margin-right: 6px;"
          />
          BookMarker
        </a>
        <div
          class="MuiBox-root css-i9gxme"
        />
        <div
          class="MuiBox-root css-5nwj3y"
        >
          <button
            aria-controls="user-menu"
            aria-expanded="true"
            aria-haspopup="true"
            aria-label="open-menu"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-gdplcp-MuiButtonBase-root-MuiIconButton-root"
            data-mui-internal-clone-element="true"
            tabindex="0"
            type="button"
          >
            <div
              class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-bwsr6u-MuiAvatar-root"
            >
              T
            </div>
          </button>
        </div>
      </div>
    </header>
  </div>
</DocumentFragment>
`;

exports[`matches with snapshot when logged out 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-i9gxme"
  >
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-1fha3o-MuiPaper-root-MuiAppBar-root"
      style="--Paper-shadow: 0px 2px 4px -1px rgba(0,0,0,0.2),0px 4px 5px 0px rgba(0,0,0,0.14),0px 1px 10px 0px rgba(0,0,0,0.12);"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-18r5587-MuiToolbar-root"
      >
        <a
          aria-label="Navigate to homepage"
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1k5ggxq-MuiButtonBase-root-MuiButton-root"
          href="/"
          tabindex="0"
        >
          <i
            class="fa-solid fa-feather"
            style="margin-right: 6px;"
          />
          BookMarker
        </a>
        <div
          class="MuiBox-root css-i9gxme"
        />
        <a
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-kuoyca-MuiButtonBase-root-MuiButton-root"
          href="/login"
          tabindex="0"
        >
          Login
        </a>
        <a
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-6zpu25-MuiButtonBase-root-MuiButton-root"
          href="/signup"
          tabindex="0"
        >
          Sign Up
        </a>
      </div>
    </header>
  </div>
</DocumentFragment>
`;
